File: server/src/services/requirement.service.ts
import { Types } from 'mongoose';
import { ApiError } from '../utils/ApiError.js';
import { ProjectModel } from '../models/Project.model.js';
import { WorkspaceMemberModel } from '../models/WorkspaceMember.model.js';
import { RequirementModel } from '../models/Requirement.model.js';
import { RequirementVersionModel } from '../models/RequirementVersion.model.js';
import { logActivity } from './activity.service.js';
import { extractRequirementsFromText } from './requirementExtraction.service.js';

const ensureProjectAccess = async (projectId: string, userId: string) => {
  if (!Types.ObjectId.isValid(projectId)) {
    throw new ApiError(400, 'Invalid project id');
  }

  const project = await ProjectModel.findById(projectId);
  if (!project) {
    throw new ApiError(404, 'Project not found');
  }

  const member = await WorkspaceMemberModel.findOne({ workspace: project.workspace, user: userId });
  if (!member) {
    throw new ApiError(403, 'You do not have access to this project');
  }

  return project;
};

const ensureRequirementAccess = async (requirementId: string, userId: string) => {
  if (!Types.ObjectId.isValid(requirementId)) {
    throw new ApiError(400, 'Invalid requirement id');
  }

  const requirement = await RequirementModel.findById(requirementId);
  if (!requirement) {
    throw new ApiError(404, 'Requirement not found');
  }

  const member = await WorkspaceMemberModel.findOne({ workspace: requirement.workspace, user: userId });
  if (!member) {
    throw new ApiError(403, 'You do not have access to this requirement');
  }

  return requirement;
};

const populateRequirement = (query: ReturnType<typeof RequirementModel.findById>) =>
  query
    .populate('project', 'name clientName status priority originalScope')
    .populate('workspace', 'name slug')
    .populate('createdBy', 'name email')
    .populate('updatedBy', 'name email');

type RequirementSnapshotSource = {
  _id: { toString: () => string };
  title: string;
  description: string;
  type: 'functional' | 'non_functional' | 'business' | 'technical' | 'ui_ux' | 'security' | 'performance' | 'integration' | 'other';
  priority: 'low' | 'medium' | 'high' | 'critical';
  status: 'proposed' | 'approved' | 'in_progress' | 'completed' | 'rejected' | 'changed';
  source: 'original_scope' | 'manual' | 'client_message' | 'meeting_note' | 'document' | 'ai_extracted';
  acceptanceCriteria: string[];
  tags: string[];
  estimatedEffort?: number | null;
};

const toRequirementSnapshot = (requirement: RequirementSnapshotSource) => ({
  requirementId: requirement._id.toString(),
  title: requirement.title,
  description: requirement.description,
  type: requirement.type,
  priority: requirement.priority,
  status: requirement.status,
  source: requirement.source,
  acceptanceCriteria: [...requirement.acceptanceCriteria],
  tags: [...requirement.tags],
  estimatedEffort: requirement.estimatedEffort,
});

export const getProjectRequirements = async (projectId: string, userId: string) => {
  await ensureProjectAccess(projectId, userId);

  const requirements = await RequirementModel.find({ project: projectId })
    .sort({ isBaseline: -1, baselineVersion: -1, createdAt: -1 })
    .populate('project', 'name clientName status priority originalScope')
    .populate('workspace', 'name slug')
    .populate('createdBy', 'name email')
    .populate('updatedBy', 'name email');

  return requirements;
};

export const getRequirementById = async (requirementId: string, userId: string) => {
  const requirement = await ensureRequirementAccess(requirementId, userId);
  return populateRequirement(RequirementModel.findById(requirement._id));
};

export const createRequirement = async (userId: string, data: {
  projectId: string;
  workspaceId?: string;
  title: string;
  description: string;
  type?: 'functional' | 'non_functional' | 'business' | 'technical' | 'ui_ux' | 'security' | 'performance' | 'integration' | 'other';
  priority?: 'low' | 'medium' | 'high' | 'critical';
  status?: 'proposed' | 'approved' | 'in_progress' | 'completed' | 'rejected' | 'changed';
  source?: 'original_scope' | 'manual' | 'client_message' | 'meeting_note' | 'document' | 'ai_extracted';
  sourceText?: string;
  acceptanceCriteria?: string[];
  tags?: string[];
  estimatedEffort?: number;
}) => {
  const project = await ensureProjectAccess(data.projectId, userId);

  const requirement = await RequirementModel.create({
    project: project._id,
    workspace: project.workspace,
    title: data.title,
    description: data.description,
    type: data.type ?? 'functional',
    priority: data.priority ?? 'medium',
    status: data.status ?? 'proposed',
    source: data.source ?? 'manual',
    sourceText: data.sourceText ?? '',
    acceptanceCriteria: data.acceptanceCriteria ?? [],
    tags: data.tags ?? [],
    estimatedEffort: data.estimatedEffort,
    createdBy: userId,
  });

  await logActivity({
    workspace: project.workspace.toString(),
    user: userId,
    action: 'REQUIREMENT_CREATED',
    entityType: 'Requirement',
    entityId: requirement._id.toString(),
    metadata: {
      requirementTitle: requirement.title,
      projectId: project._id.toString(),
      projectName: project.name,
    },
  });

  return populateRequirement(RequirementModel.findById(requirement._id));
};

export const updateRequirement = async (requirementId: string, userId: string, data: {
  title?: string;
  description?: string;
  type?: 'functional' | 'non_functional' | 'business' | 'technical' | 'ui_ux' | 'security' | 'performance' | 'integration' | 'other';
  priority?: 'low' | 'medium' | 'high' | 'critical';
  status?: 'proposed' | 'approved' | 'in_progress' | 'completed' | 'rejected' | 'changed';
  source?: 'original_scope' | 'manual' | 'client_message' | 'meeting_note' | 'document' | 'ai_extracted';
  sourceText?: string;
  acceptanceCriteria?: string[];
  tags?: string[];
  estimatedEffort?: number;
}) => {
  const requirement = await ensureRequirementAccess(requirementId, userId);

  if (data.title !== undefined) requirement.title = data.title;
  if (data.description !== undefined) requirement.description = data.description;
  if (data.type !== undefined) requirement.type = data.type;
  if (data.priority !== undefined) requirement.priority = data.priority;
  if (data.status !== undefined) requirement.status = data.status;
  if (data.source !== undefined) requirement.source = data.source;
  if (data.sourceText !== undefined) requirement.sourceText = data.sourceText;
  if (data.acceptanceCriteria !== undefined) requirement.acceptanceCriteria = data.acceptanceCriteria;
  if (data.tags !== undefined) requirement.tags = data.tags;
  if (data.estimatedEffort !== undefined) requirement.estimatedEffort = data.estimatedEffort;
  requirement.updatedBy = new Types.ObjectId(userId);

  await requirement.save();

  await logActivity({
    workspace: requirement.workspace.toString(),
    user: userId,
    action: 'REQUIREMENT_UPDATED',
    entityType: 'Requirement',
    entityId: requirement._id.toString(),
    metadata: {
      requirementTitle: requirement.title,
      projectId: requirement.project.toString(),
      requirementId: requirement._id.toString(),
    },
  });

  return populateRequirement(RequirementModel.findById(requirement._id));
};

export const deleteRequirement = async (requirementId: string, userId: string) => {
  const requirement = await ensureRequirementAccess(requirementId, userId);

  await RequirementModel.findByIdAndDelete(requirementId);

  await logActivity({
    workspace: requirement.workspace.toString(),
    user: userId,
    action: 'REQUIREMENT_DELETED',
    entityType: 'Requirement',
    entityId: requirementId,
    metadata: {
      requirementTitle: requirement.title,
      projectId: requirement.project.toString(),
    },
  });

  return requirement;
};

export const extractRequirements = async (userId: string, data: {
  projectId: string;
  sourceText: string;
  source?: 'original_scope' | 'manual' | 'client_message' | 'meeting_note' | 'document' | 'ai_extracted';
}) => {
  const project = await ensureProjectAccess(data.projectId, userId);
  const suggestions = extractRequirementsFromText(data.sourceText, data.source ?? 'original_scope');

  await logActivity({
    workspace: project.workspace.toString(),
    user: userId,
    action: 'REQUIREMENTS_EXTRACTED',
    entityType: 'Project',
    entityId: project._id.toString(),
    metadata: {
      projectId: project._id.toString(),
      projectName: project.name,
      source: data.source ?? 'original_scope',
      extractedCount: suggestions.length,
    },
  });

  return suggestions;
};

export const createRequirementBaseline = async (userId: string, data: {
  projectId: string;
  label?: string;
  description?: string;
}) => {
  const project = await ensureProjectAccess(data.projectId, userId);
  const requirements = await RequirementModel.find({ project: project._id }).sort({ createdAt: 1 });

  if (requirements.length === 0) {
    throw new ApiError(400, 'Create at least one requirement before building a baseline');
  }

  const latestVersion = await RequirementVersionModel.findOne({ project: project._id }).sort({ versionNumber: -1 });
  const versionNumber = (latestVersion?.versionNumber ?? 0) + 1;

  const version = await RequirementVersionModel.create({
    project: project._id,
    workspace: project.workspace,
    versionNumber,
    label: data.label?.trim() || (versionNumber === 1 ? 'Initial Baseline' : 'Manual Update'),
    description: data.description?.trim() ?? '',
      requirementsSnapshot: requirements.map((requirement) => toRequirementSnapshot(requirement)),
    createdBy: userId,
  });

  await RequirementModel.updateMany(
    { project: project._id },
    { $set: { isBaseline: true, baselineVersion: versionNumber } }
  );

  await logActivity({
    workspace: project.workspace.toString(),
    user: userId,
    action: 'REQUIREMENT_BASELINE_CREATED',
    entityType: 'RequirementVersion',
    entityId: version._id.toString(),
    metadata: {
      projectId: project._id.toString(),
      projectName: project.name,
      versionNumber,
      requirementCount: requirements.length,
    },
  });

  return RequirementVersionModel.findById(version._id).populate('createdBy', 'name email');
};

export const getRequirementVersions = async (projectId: string, userId: string) => {
  await ensureProjectAccess(projectId, userId);

  return RequirementVersionModel.find({ project: projectId })
    .sort({ versionNumber: -1 })
    .populate('createdBy', 'name email');
};

File: server-go/internal/modules/requirement/requirement_service.go
package requirement

import (
	"context"
	"errors"
	"fmt"
	"time"

	"driftledger/server-go/internal/modules/activity"
	"driftledger/server-go/internal/utils"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type Service struct{ db *mongo.Database }

func NewService(db *mongo.Database) Service { return Service{db: db} }

func (s Service) Create(ctx context.Context, userID primitive.ObjectID, p CreateRequirementRequest) (Requirement, error) {
	projectID, err := utils.ObjectID(p.ProjectID)
	if err != nil {
		return Requirement{}, err
	}
	project, err := utils.RequireProjectAccess(ctx, s.db, projectID, userID)
	if err != nil {
		return Requirement{}, err
	}
	if p.WorkspaceID != "" && p.WorkspaceID != project.Workspace.Hex() {
		return Requirement{}, utils.ErrForbidden
	}
	now := time.Now().UTC()
	req := Requirement{ID: utils.NewID(), Project: projectID, Workspace: project.Workspace, Title: p.Title, Description: p.Description, Type: def(p.Type, "functional"), Priority: def(p.Priority, "medium"), Status: def(p.Status, "proposed"), Source: def(p.Source, "manual"), SourceText: p.SourceText, AcceptanceCriteria: p.AcceptanceCriteria, Tags: p.Tags, EstimatedEffort: p.EstimatedEffort, IsBaseline: false, BaselineVersion: 0, CreatedBy: userID, UpdatedBy: userID, CreatedAt: now, UpdatedAt: now}
	_, err = s.db.Collection("requirements").InsertOne(ctx, req)
	if err == nil {
		activity.Log(ctx, s.db, project.Workspace, userID, "REQUIREMENT_CREATED", "Requirement", req.ID.Hex(), bson.M{"title": req.Title, "projectId": projectID.Hex()})
	}
	return req, err
}

func (s Service) ListByProject(ctx context.Context, projectID, userID primitive.ObjectID) ([]Requirement, error) {
	if _, err := utils.RequireProjectAccess(ctx, s.db, projectID, userID); err != nil {
		return nil, err
	}
	cursor, err := s.db.Collection("requirements").Find(ctx, bson.M{"project": projectID}, options.Find().SetSort(bson.D{{Key: "createdAt", Value: -1}}))
	if err != nil {
		return nil, err
	}
	var reqs []Requirement
	err = cursor.All(ctx, &reqs)
	return reqs, err
}

func (s Service) Get(ctx context.Context, id, userID primitive.ObjectID) (Requirement, error) {
	var req Requirement
	err := s.db.Collection("requirements").FindOne(ctx, bson.M{"_id": id}).Decode(&req)
	if errors.Is(err, mongo.ErrNoDocuments) {
		return req, utils.ErrNotFound
	}
	if err != nil {
		return req, err
	}
	_, err = utils.RequireProjectAccess(ctx, s.db, req.Project, userID)
	return req, err
}

func (s Service) Update(ctx context.Context, id, userID primitive.ObjectID, p UpdateRequirementRequest) (Requirement, error) {
	req, err := s.Get(ctx, id, userID)
	if err != nil {
		return req, err
	}
	update := bson.M{"updatedAt": time.Now().UTC(), "updatedBy": userID}
	if p.Title != "" {
		update["title"] = p.Title
	}
	if p.Description != "" {
		update["description"] = p.Description
	}
	if p.Type != "" {
		update["type"] = p.Type
	}
	if p.Priority != "" {
		update["priority"] = p.Priority
	}
	if p.Status != "" {
		update["status"] = p.Status
	}
	if p.Source != "" {
		update["source"] = p.Source
	}
	if p.SourceText != "" {
		update["sourceText"] = p.SourceText
	}
	if p.AcceptanceCriteria != nil {
		update["acceptanceCriteria"] = p.AcceptanceCriteria
	}
	if p.Tags != nil {
		update["tags"] = p.Tags
	}
	if p.EstimatedEffort != nil {
		update["estimatedEffort"] = p.EstimatedEffort
	}
	_, err = s.db.Collection("requirements").UpdateOne(ctx, bson.M{"_id": id}, bson.M{"$set": update})
	if err == nil {
		activity.Log(ctx, s.db, req.Workspace, userID, "REQUIREMENT_UPDATED", "Requirement", id.Hex(), bson.M{"title": req.Title})
	}
	return s.Get(ctx, id, userID)
}

func (s Service) Delete(ctx context.Context, id, userID primitive.ObjectID) error {
	req, err := s.Get(ctx, id, userID)
	if err != nil {
		return err
	}
	_, err = s.db.Collection("requirements").DeleteOne(ctx, bson.M{"_id": id})
	if err == nil {
		activity.Log(ctx, s.db, req.Workspace, userID, "REQUIREMENT_DELETED", "Requirement", id.Hex(), bson.M{"title": req.Title})
	}
	return err
}

func (s Service) Baseline(ctx context.Context, userID primitive.ObjectID, p BaselineRequest) (RequirementVersion, error) {
	projectID, err := utils.ObjectID(p.ProjectID)
	if err != nil {
		return RequirementVersion{}, err
	}
	project, err := utils.RequireProjectAccess(ctx, s.db, projectID, userID)
	if err != nil {
		return RequirementVersion{}, err
	}
	reqs, err := s.ListByProject(ctx, projectID, userID)
	if err != nil {
		return RequirementVersion{}, err
	}
	if len(reqs) == 0 {
		return RequirementVersion{}, errors.New("empty baseline")
	}
	var latest RequirementVersion
	_ = s.db.Collection("requirementversions").FindOne(ctx, bson.M{"project": projectID}, options.FindOne().SetSort(bson.D{{Key: "versionNumber", Value: -1}})).Decode(&latest)
	versionNumber := latest.VersionNumber + 1
	snapshots := make([]RequirementSnapshot, 0, len(reqs))
	for _, req := range reqs {
		snapshots = append(snapshots, RequirementSnapshot{RequirementID: req.ID.Hex(), Title: req.Title, Description: req.Description, Type: req.Type, Priority: req.Priority, Status: req.Status, Source: req.Source, AcceptanceCriteria: req.AcceptanceCriteria, Tags: req.Tags, EstimatedEffort: req.EstimatedEffort})
	}
	if p.Label == "" {
		p.Label = fmt.Sprintf("Baseline v%d", versionNumber)
	}
	if p.Description == "" {
		p.Description = fmt.Sprintf("Requirement baseline version %d", versionNumber)
	}
	version := RequirementVersion{ID: utils.NewID(), Project: projectID, Workspace: project.Workspace, VersionNumber: versionNumber, Label: p.Label, Description: p.Description, RequirementsSnapshot: snapshots, CreatedBy: userID, CreatedAt: time.Now().UTC()}
	_, err = s.db.Collection("requirementversions").InsertOne(ctx, version)
	if err == nil {
		_, _ = s.db.Collection("requirements").UpdateMany(ctx, bson.M{"project": projectID}, bson.M{"$set": bson.M{"isBaseline": true, "baselineVersion": versionNumber, "updatedAt": time.Now().UTC(), "updatedBy": userID}})
		activity.Log(ctx, s.db, project.Workspace, userID, "REQUIREMENT_BASELINE_CREATED", "RequirementVersion", version.ID.Hex(), bson.M{"projectId": projectID.Hex(), "versionNumber": versionNumber})
	}
	return version, err
}

func (s Service) Versions(ctx context.Context, projectID, userID primitive.ObjectID) ([]RequirementVersion, error) {
	if _, err := utils.RequireProjectAccess(ctx, s.db, projectID, userID); err != nil {
		return nil, err
	}
	cursor, err := s.db.Collection("requirementversions").Find(ctx, bson.M{"project": projectID}, options.Find().SetSort(bson.D{{Key: "versionNumber", Value: -1}}))
	if err != nil {
		return nil, err
	}
	var versions []RequirementVersion
	err = cursor.All(ctx, &versions)
	return versions, err
}

func def(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
